pkg/search/postgres/store.go

```diff
@@ -0,0 +1,117 @@
+
+// GenericStore implements subset of Store interface for resources with single ID.
+type GenericStore[T any, PT unmarshaler[T]] struct {
+	db                               postgres.DB
+	targetResource                   permissions.ResourceMetadata
+	schema                           *walker.Schema
+	setPostgresOperationDurationTime durationTimeSetter
+	setAcquireDBConnDuration         durationTimeSetter
+	permissionChecker                PermissionChecker
+	pkGetter                         primaryKeyGetter[T, PT]
+}
+
+// NewGenericStore returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStore[T any, PT unmarshaler[T]](
+	db postgres.DB,
+	targetResource permissions.ResourceMetadata,
+	schema *walker.Schema,
+	setPostgresOperationDurationTime durationTimeSetter,
+	setAcquireDBConnDuration durationTimeSetter,
+	pkGetter primaryKeyGetter[T, PT],
+) *GenericStore[T, PT] {
+	return &GenericStore[T, PT]{
+		db:                               db,
+		targetResource:                   targetResource,
+		schema:                           schema,
+		setPostgresOperationDurationTime: setPostgresOperationDurationTime,
+		setAcquireDBConnDuration:         setAcquireDBConnDuration,
+		pkGetter:                         pkGetter,
+	}
+}
+
+// NewGenericStoreWithPermissionChecker returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStoreWithPermissionChecker[T any, PT unmarshaler[T]](
+	db postgres.DB,
+	checker PermissionChecker,
+	schema *walker.Schema,
+	setPostgresOperationDurationTime durationTimeSetter,
+	setAcquireDBConnDuration durationTimeSetter,
+	pkGetter primaryKeyGetter[T, PT],
+) *GenericStore[T, PT] {
+	return &GenericStore[T, PT]{
+		db:                               db,
+		schema:                           schema,
+		setPostgresOperationDurationTime: setPostgresOperationDurationTime,
+		setAcquireDBConnDuration:         setAcquireDBConnDuration,
+		permissionChecker:                checker,
+		pkGetter:                         pkGetter,
+	}
+}
+
+func (s *GenericStore[T, PT]) hasPermissionsChecker() bool {
+	return s.permissionChecker != nil
+}
+
+// AcquireConn returns Acquires new connection from DB.
+func (s *GenericStore[T, PT]) AcquireConn(ctx context.Context, op ops.Op) (*postgres.Conn, error) {
+	defer s.setAcquireDBConnDuration(time.Now(), op)
+	conn, err := s.db.Acquire(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// Exists returns if the ID exists in the store.
+func (s *GenericStore[T, PT]) Exists(ctx context.Context, id string) (bool, error) {
+	defer s.setPostgresOperationDurationTime(time.Now(), ops.Exists)
+
+	var sacQueryFilter *v1.Query
+	if s.hasPermissionsChecker() {
+		if ok, err := s.permissionChecker.ExistsAllowed(ctx); err != nil {
+			return false, err
+		} else if !ok {
+			return false, sac.ErrResourceAccessDenied
```

Comment:
This changes the behaviour compared to the generated code.
I'd stick to the previous behaviour that returns false, nil, as if the item did not exist, rather than an error that already tells that the item exists. I think from a security point of view, the new behaviour is a data leak.